### PR TITLE
plugin/forward: gracefull stop

### PR DESCRIPTION
**Pull request by rdrozhdzh**
### 1. What does this pull request do?
stop connection manager only when no queries in progress

### 2. Which issues (if any) are related?
#1666

### 3. Which documentation changes (if any) need to be made?
none

### CODE REVIEW

File: plugin/forward/connect.go [hunk 1]
```diff
@@ -34,6 +34,16 @@ func (p *Proxy) updateRtt(newRtt time.Duration) {
 }
 
 func (p *Proxy) connect(ctx context.Context, state request.Request, forceTCP, metric bool) (*dns.Msg, error) {
+	atomic.AddInt32(&p.inProgress, 1)
+	defer func() {
+		if atomic.AddInt32(&p.inProgress, -1) == 0 {
```
Comment (miekg) [inline plugin/forward/connect.go:40]: how is guaranteed to not hit zero while the proxy is still up and running? (i.e. we are not shutting down)

Comment (rdrozhdzh) [reply plugin/forward/connect.go:40]: it is ok to have zero here, method checkStopTransport() will not stop transport unless Proxy.state == stopping

File: plugin/forward/persistent.go [hunk 2]
```diff
@@ -157,7 +157,18 @@ func (t *transport) Yield(c *dns.Conn) {
 }
 
 // Stop stops the transport's connection manager.
-func (t *transport) Stop() { t.stop <- true }
+func (t *transport) Stop() {
+	t.stop <- true
+
+	// at this point nothing should write to channels anymore
+	// closing channels to make tests panic on write
```
Comment (miekg) [inline plugin/forward/persistent.go]: I find this `close`-list somewhat ugly, is this only needed for the test? If so can we just do it in the tests only?

**Review by miekg — COMMENTED** [review]
overall lgtm

File: plugin/forward/persistent.go (hunk 2 above)
Comment (rdrozhdzh) [reply plugin/forward/persistent.go]: Ok, I will remove this.
The idea was is that panic on writing in closed channel is easier to detect and fix than hanging if nobody reads the channel

**Comment by rdrozhdzh** [comment]
@miekg, the `proxy_test.go` has yet another reference to `golang.org/x/net/context`. After merge, this likely should be changed in PR #1699 as well. Or I can change it here, if you are going to merge 1699 first

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>@miekg, the `proxy_test.go` has yet another reference to `golang.org/x/net/context`. After merge, this likely should be changed in PR #1699 as well. Or I can change it here, if you are going to merge 1699 first

Merged 1699

**Review by miekg — APPROVED** [review]

**Pull request merged**